### PR TITLE
add minimum cmake support

### PR DESCRIPTION
**Pull request by cyjseagull**
add minimum cmake support
1. set compiler settings (EthCompilerSettings.cmake)
2. set compiling options (EthOptions.cmake)
3. install dependencies according to inner cmake into "deps" directory (boost, jsoncpp, jsonrpccpp, crypto, secp256k1)
4. generate BuildInfo.h with EthBuildInfo.cmake (taken cmake/templates/BuildInfo.h.in as template)
5. add build.sh to setup and compile lab-bcos easily
<a href='#crh-start'></a><a href='#crh-data-%7B%22approved%22%3A%20%7B%22https%3A//github.com/cyjseagull%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/4343461%3Fv%3D4%22%7D%2C%20%22https%3A//github.com/wheatli%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/23206367%3Fv%3D4%22%7D%7D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23issuecomment-414677253%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%3Ashipit%3A%22%2C%20%22created_at%22%3A%20%222018-08-21T13%3A39%3A45Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/23206367%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/wheatli%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3Ashipit%3A%22%2C%20%22created_at%22%3A%20%222018-08-21T13%3A41%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/4343461%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cyjseagull%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%206b09ae1946e42186495999229f5bde4cf2d26770%20cmake/EthOptions.cmake%2044%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23discussion_r211583439%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%23ARCH%20TYPE%20should%20be%20putted%20to%20near%20ARCH%20TYPE%22%2C%20%22created_at%22%3A%20%222018-08-21T12%3A21%3A36Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/23206367%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/wheatli%22%7D%7D%2C%20%7B%22body%22%3A%20%22done%22%2C%20%22created_at%22%3A%20%222018-08-21T13%3A37%3A17Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/4343461%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cyjseagull%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222018-08-21T13%3A38%3A55Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/4343461%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cyjseagull%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20cmake/EthOptions.cmake%3AL1-94%22%7D%2C%20%22Pull%206b09ae1946e42186495999229f5bde4cf2d26770%20cmake/EthOptions.cmake%2048%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23discussion_r211582755%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22default%20setted%20as%20ON%22%2C%20%22created_at%22%3A%20%222018-08-21T12%3A19%3A01Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/23206367%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/wheatli%22%7D%7D%2C%20%7B%22body%22%3A%20%22done%22%2C%20%22created_at%22%3A%20%222018-08-21T13%3A37%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/4343461%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cyjseagull%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222018-08-21T13%3A38%3A56Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/4343461%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cyjseagull%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20cmake/EthOptions.cmake%3AL1-94%22%7D%2C%20%22Pull%206b09ae1946e42186495999229f5bde4cf2d26770%20tools/build.sh%2072%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23discussion_r211581734%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22We%20need%20to%20check%20wether%20it%20is%20CentOS%5Cuff0c%20log%20error%20if%20not%22%2C%20%22created_at%22%3A%20%222018-08-21T12%3A15%3A19Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/23206367%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/wheatli%22%7D%7D%2C%20%7B%22body%22%3A%20%22already%20resolved%2C%20add%20%5C%22get_platform%5C%22%20function%20to%20obtain%20platform%20type%3A%5Cr%5Cncommit%20url%5Cuff1ahttps%3A//github.com/FISCO-BCOS/lab-bcos/pull/2/commits/62ef3f2924f67b3e261ea7750c6b7042b50d45e9%22%2C%20%22created_at%22%3A%20%222018-08-21T13%3A28%3A04Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/4343461%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cyjseagull%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222018-08-21T13%3A38%3A57Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/4343461%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cyjseagull%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20tools/build.sh%3AL1-102%22%7D%2C%20%22Pull%206b09ae1946e42186495999229f5bde4cf2d26770%20tools/build.sh%2095%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23discussion_r211581933%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22need%20to%20check%20CentOS%20too%2C%20log%20error%20if%20not%22%2C%20%22created_at%22%3A%20%222018-08-21T12%3A16%3A08Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/23206367%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/wheatli%22%7D%7D%2C%20%7B%22body%22%3A%20%22done%22%2C%20%22created_at%22%3A%20%222018-08-21T13%3A36%3A54Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/4343461%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cyjseagull%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-complete%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222018-08-21T13%3A38%3A57Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/4343461%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cyjseagull%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20tools/build.sh%3AL1-102%22%7D%7D%2C%20%22processed%22%3A%20%5B%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23issuecomment-414677253%22%2C%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23issuecomment-414677990%22%2C%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23discussion_r211581734%22%2C%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23discussion_r211581933%22%2C%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23discussion_r211582755%22%2C%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23discussion_r211583439%22%2C%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23discussion_r211604291%22%2C%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23discussion_r211607423%22%2C%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23discussion_r211607484%22%2C%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23discussion_r211607541%22%2C%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23discussion_r211608139%22%2C%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23discussion_r211608148%22%2C%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23discussion_r211608152%22%2C%20%22https%3A//github.com/FISCO-BCOS/lab-bcos/pull/2%23discussion_r211608155%22%5D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>

<a href='https://github.com/cyjseagull'></a><a href='https://github.com/wheatli'></a>

- [x] <a href='#crh-comment-Pull 6b09ae1946e42186495999229f5bde4cf2d26770 tools/build.sh 72'></a> &nbsp;<b><a href='https://github.com/FISCO-BCOS/lab-bcos/pull/2#discussion_r211581734'>File: tools/build.sh:L1-102</a></b>
- <a href='https://github.com/wheatli'></a> We need to check wether it is CentOS， log error if not
- <a href='https://github.com/cyjseagull'></a> already resolved, add "get_platform" function to obtain platform type:
commit url：https://github.com/FISCO-BCOS/lab-bcos/pull/2/commits/62ef3f2924f67b3e261ea7750c6b7042b50d45e9
- [x] <a href='#crh-comment-Pull 6b09ae1946e42186495999229f5bde4cf2d26770 tools/build.sh 95'></a> &nbsp;<b><a href='https://github.com/FISCO-BCOS/lab-bcos/pull/2#discussion_r211581933'>File: tools/build.sh:L1-102</a></b>
- <a href='https://github.com/wheatli'></a> need to check CentOS too, log error if not
- <a href='https://github.com/cyjseagull'></a> done
- [x] <a href='#crh-comment-Pull 6b09ae1946e42186495999229f5bde4cf2d26770 cmake/EthOptions.cmake 48'></a> &nbsp;<b><a href='https://github.com/FISCO-BCOS/lab-bcos/pull/2#discussion_r211582755'>File: cmake/EthOptions.cmake:L1-94</a></b>
- <a href='https://github.com/wheatli'></a> default setted as ON
- <a href='https://github.com/cyjseagull'></a> done
- [x] <a href='#crh-comment-Pull 6b09ae1946e42186495999229f5bde4cf2d26770 cmake/EthOptions.cmake 44'></a> &nbsp;<b><a href='https://github.com/FISCO-BCOS/lab-bcos/pull/2#discussion_r211583439'>File: cmake/EthOptions.cmake:L1-94</a></b>
- <a href='https://github.com/wheatli'></a> #ARCH TYPE should be putted to near ARCH TYPE
- <a href='https://github.com/cyjseagull'></a> done

<a href='https://www.codereviewhub.com/FISCO-BCOS/lab-bcos/pull/2?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/FISCO-BCOS/lab-bcos/pull/2?approve=1'></a>&nbsp;<a href='https://www.codereviewhub.com/FISCO-BCOS/lab-bcos/pull/2?approve=0'></a>&nbsp;<a href='https://github.com/FISCO-BCOS/lab-bcos/pull/2'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: tools/build.sh [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+install_centos_deps() {
+    install_centos_package "cmake3" "gcc-c++" "openssl" "openssl-devel"
+}
+
+#=== install all deps
+install_all_deps()
+{
+    if grep -Eqi "Ubuntu" /etc/issue || grep -Eq "Ubuntu" /etc/*-release; then
+        install_ubuntu_deps
+    else
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply tools/build.sh]: 已解决，加了get_platform函数获取平台类型：
commit连接：https://github.com/FISCO-BCOS/lab-bcos/pull/2/commits/62ef3f2924f67b3e261ea7750c6b7042b50d45e9

File: tools/build.sh [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    execute_cmd "mkdir -p build && cd build/"
+    execute_cmd "cmake3 .. "
+    execute_cmd "make -j2 && make install && cd ${current_dir}"
+}
+
+build_source() {
+    cd ${current_dir}
+    if grep -Eqi "Ubuntu" /etc/issue || grep -Eq "Ubuntu" /etc/*-release; then
+        build_ubuntu_source
+    else
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply tools/build.sh]: done

File: cmake/EthOptions.cmake [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            "Choose the type of build, options are: Debug Release RelWithDebInfo MinSizeRel." FORCE)
+    endif()
+
+    eth_default_option(BUILD_SHARED_LIBS OFF)
+   
+    #ARCH TYPE
+    eth_default_option(STATIC_BUILD OFF)
+    eth_default_option(ARCH_TYPE OFF)
+    # unit tests
+    eth_default_option(TESTS OFF)
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply cmake/EthOptions.cmake]: done

File: cmake/EthOptions.cmake [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    # Default to RelWithDebInfo configuration if no configuration is explicitly specified.
+    if (NOT CMAKE_BUILD_TYPE)
+        set(CMAKE_BUILD_TYPE "RelWithDebInfo" CACHE STRING
+            "Choose the type of build, options are: Debug Release RelWithDebInfo MinSizeRel." FORCE)
+    endif()
+
+    eth_default_option(BUILD_SHARED_LIBS OFF)
+   
+    #ARCH TYPE
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply cmake/EthOptions.cmake]: done

Comment (cyjseagull) [inline cmake/EthOptions.cmake]: :+1:<a href='#crh-update-none'></a>

File: cmake/EthOptions.cmake (hunk 3 above)
Comment (cyjseagull) [inline cmake/EthOptions.cmake]: :+1:<a href='#crh-update-none'></a>

File: tools/build.sh (hunk 1 above)
Comment (cyjseagull) [inline tools/build.sh]: :+1:<a href='#crh-update-none'></a>

File: tools/build.sh (hunk 2 above)
Comment (cyjseagull) [inline tools/build.sh]: :+1:<a href='#crh-update-complete'></a>

**Comment by wheatli** [comment]
:shipit:

**Comment by cyjseagull** [comment]
:shipit:

**Pull request merged**